lib/zstd/src/mutex.cpp

```diff
@@ -0,0 +1,276 @@
+#ifdef CONFIG_ZSTD_MUTEX
+#include <zstd/mutex.h>
+
+#include <exception>
+#include <system_error>
+
+namespace zstd
+{
+
+using std::system_error;
+using std::bad_alloc;
+using std::system_category;
+
+/*******************************************************************************
+ * Slab based new/delete?
+ ******************************************************************************/
+#ifdef CONFIG_ZSTD_NEW_SLAB
+
+#define MAX(a, b)      ((a > b) ? a : b)
+#define MAX_SIZE(a, b) MAX(sizeof(a), sizeof(b))
+
+#define MAX_MUTEX_SIZE_NON_TIMED MAX_SIZE(mutex, recursive_mutex)
+#define MAX_MUTEX_SIZE_TIMED     MAX_SIZE(timed_mutex, recursive_timed_mutex)
+
+#define MAX_MUTEX_SIZE MAX(MAX_MUTEX_SIZE_NON_TIMED, MAX_MUTEX_SIZE_TIMED)
+
+static K_MEM_SLAB_DEFINE(mutex_slab, MAX_MUTEX_SIZE, \
+                  CONFIG_ZSTD_MUTEX_NEW_SLAB_MAX_COUNT, 1);
+
+#endif
+
+/*******************************************************************************
+ * zstd::mutex
+ ******************************************************************************/
+mutex::mutex() : m_mtx_ptr{&m_mtx}
+{
+	k_mutex_init(&m_mtx);
+}
+
+#ifdef CONFIG_ZSTD_NEW_HEAP
+void *mutex::operator new(size_t size)
+{
+	void *ptr = k_calloc(1, size);
+
+	if (ptr == nullptr) {
+		throw bad_alloc();
+	}
+
+	return ptr;
+}
+
+void mutex::operator delete(void *ptr)
+{
+	free(ptr);
+}
+#elif CONFIG_ZSTD_NEW_SLAB
+void *mutex::operator new(size_t size)
+{
+	void* ptr = nullptr;
+
+	int rv = k_mem_slab_alloc(&mutex_slab, &ptr, K_NO_WAIT);
+
+	if (rv != 0) {
+		throw bad_alloc();
+	}
+
+	return ptr;
+}
+
+void mutex::operator delete(void *ptr)
+{
+	k_mem_slab_free(&mutex_slab, &ptr);
+}
+#endif
+
+void mutex::lock()
+{
+	if ((m_mtx_ptr->lock_count > 0) &&
+		(k_current_get() == m_mtx_ptr->owner)) {
+		throw system_error(-1, system_category());
+	}
+
+	int rv = k_mutex_lock(m_mtx_ptr, K_FOREVER);
+
+	if (rv != 0) {
+		throw system_error(rv, system_category());
+	}
+}
+
+bool mutex::try_lock()
+{
+	if ((m_mtx_ptr->lock_count > 0) &&
+		(k_current_get() == m_mtx_ptr->owner)) {
+		return false;
```

Comment:
This function would cause deadlock if it were not for this check so the correct exception + code should be thrown.